haskell_scotty/Server.hs

```diff
@@ -22,4 +22,9 @@ main = scotty 6969 $ do
     jsonAPI client
     status status201
 
+  delete "/api/clients/:id" $ do
+    clientId <- param "id"
+    liftIO $ deleteClient clientId
```

Comment:
`liftIO` because the `do` block of the delete action is run in the `Action` monad, not in the `IO` monad, while `deleteClient`  wants to be run in the `IO` monad (it says so right in its type declaration). Since you've written `deleteClient`, you could change it to be an `Action` and do `liftIO` inside of it for the things that want to be in `IO`, but I think what you have is better.

I would never desugar any of the `do` syntax here. Are you asking about the `do` in `main` or the `do` in `deleteClient`? The `status` is the return value of the former, while the return value of the latter is just unit wrapped in IO, since every Haskell function has to return something of the appropriate type. The `execute` call actually returns an IO-wrapped `Int64` for some reason, so you couldn't just stop there or the types wouldn't match.

Note that `return` doesn't do what you might think it does: in particular it does not affect flow control in any way. It just takes a value and "wraps" it in a monad wrapper for when you need that.

I'm going to merge this code, then do a patch to turn on `-Wall` with cabal and fix all the warnings it triggers.